src/lib.rs

```diff
@@ -26,3 +28,31 @@ where
     <A as AsMut<[T]>>::as_mut(&mut a).copy_from_slice(slice);
     a
 }
+
+/// Try to fill buf with data from source, dealing with short reads such as
+/// caused by Chain.
+///
+/// Errors: See `std::io::Read`.
+fn read_bytes<R>(source: &mut R, buf: &mut [u8]) -> Result<usize>
+where
+    R: Read,
+{
+    let mut offset = 0;
+    loop {
+        match source.read(&mut buf[offset..]) {
+            Ok(0) => {
+                return Ok(offset);
+            }
+            Ok(n) => {
+                offset += n;
+                if offset == buf.len() {
+                    return Ok(offset);
+                }
+            }
+            Err(ref e) if e.kind() == ErrorKind::Interrupted => {}
```

Comment:
Why eat this error?  I know the documentation says it can be retried but should that be this libraries responsibility or the client consuming it?